src/StmtToViz.cpp

```diff
@@ -0,0 +1,2900 @@
+#include "StmtToViz.h"
+#include "Debug.h"
+#include "Error.h"
+#include "IROperator.h"
+#include "IRPrinter.h"
+#include "IRVisitor.h"
+#include "Module.h"
+#include "Scope.h"
+#include "Substitute.h"
+#include "Util.h"
+
+#include <cstdio>
+#include <fstream>
+#include <iostream>
+#include <iterator>
+#include <regex>
+#include <sstream>
+#include <utility>
+
+namespace Halide {
+namespace Internal {
+
+using std::istringstream;
+using std::ostringstream;
+using std::string;
+
+extern "C" unsigned char halide_html_template_StmtToViz_dependencies[];
+extern "C" unsigned char halide_html_template_StmtToViz_stylesheet[];
+extern "C" unsigned char halide_html_template_StmtToViz_javascript[];
+
+// Classes define within this file
+class CostModel;
+class AssemblyInfo;
+template<typename T>
+class HTMLCodePrinter;
+class HTMLVisualizationPrinter;
+class IRVisualizer;
+
+/********************** IRCostModel **********************/
+// A basic cost model for Halide IR. Estimates computation
+// cost through simple op-counting and data-movement cost
+// by counting the number of bits being moved.
+class IRCostModel : public IRVisitor {
+public:
+    IRCostModel()
+        : _max_compute_cost(-1), _max_data_cost(-1), _max_compute_cost_inclusive(-1),
+          _max_data_cost_inclusive(-1) {
+    }
+
+    // Pre-compute all costs to avoid repeated work
+    void comput_all_costs(const Module &m) {
+        // Compute all node costs
+        for (const auto &fn : m.functions()) {
+            fn.body.accept(this);
+        }
+
+        // Compute the max cost for each category
+        _max_compute_cost = -1;
+        for (auto const &entry : _compute_cost) {
+            _max_compute_cost = std::max(entry.second, _max_compute_cost);
+        }
+
+        _max_data_cost = -1;
+        for (auto const &entry : _data_cost) {
+            _max_data_cost = std::max(entry.second, _max_data_cost);
+        }
+
+        _max_compute_cost_inclusive = -1;
+        for (auto const &entry : _compute_cost_inclusive) {
+            _max_compute_cost_inclusive = std::max(entry.second, _max_compute_cost_inclusive);
+        }
+
+        _max_data_cost_inclusive = -1;
+        for (auto const &entry : _data_cost_inclusive) {
+            _max_data_cost_inclusive = std::max(entry.second, _max_data_cost_inclusive);
+        }
+    }
+
+    // Returns the compute cost of a node (estimated using simple op-counting)
+    int get_compute_cost(const IRNode *node, bool _include_subtree_cost) {
+        internal_assert(node != nullptr) << "IRCostModel::get_compute_cost(): node is nullptr\n";
+
+        int cost = -1;
+        if (_compute_cost.count(node)) {
+            cost = _include_subtree_cost ? _compute_cost_inclusive[node] : _compute_cost[node];
+        } else {
+            internal_assert(false) << "IRCostModel::get_compute_cost(): cost lookup failed\n";
```

Comment:
Instead of `internal_assert(false) << foo`, use `internal_error << foo` -- it's identical but terser and cleaner.